### DISCUSSION TITLE
Do I need `jotai-tanstack-query` ?

### DISCUSSION THREAD

**Discussion by amirhhashemi** [pmndrs/jotai#1797] [Q&A] [answered]
It's the first time I'm using `react-query` and `jotai` in the same project and I can't decide whether I need `jotai-tanstack-query` or not.

My use case is simple. I have a query that depends on the value of an atom that is created with `atomWithStorage`:

```ts
const browserIdAtom = atomWithStorage("browserId", undefined)

export const useBrowserInfo = () => {
  const [browserId] = useAtom(browserIdAtom)
  return useQuery({
    queryKey: ["browser", browserId], 
    queryFn: () => fetch(`/browser/${browserId}`)
}
```
The value of `browserIdAtom` will change based on the user's actions.

With your `react-query` integration it would be something like this:
```ts
const browserIdAtom = atomWithStorage("browserId", undefined)

const [browserAtom, browserStatusAtom] = atomsWithQuery((get) => ({
    queryKey: ["browser", get(browserIdAtom)], 
    queryFn: ({ queryKey: [, id] }) => fetch(`/browser/${id}`)
}))

export const useBrowserInfo = () => useAtom(browserAtom)
```
Is there any particular reason to prefer the second approach over the first one? I mean, they are both equal in terms of functionality, right? The only drawback I see is that I have to use `useAtom(browserStatusAtom)` to access the query status. Is there any other difference that I should be aware of?

**Comment by t6adev** [comment, marked as answer]
Don't create atoms in hooks/components without useMemo.
You can just do it simply:
```
const browserIdAtom = atomWithStorage("browserId", undefined)
const [userAtom] = atomsWithQuery((get) => ({
  queryKey: ["browser", get(browserIdAtom)], 
  queryFn: ({ queryKey: [, id] }) => fetch(`/browser/${id}`)
}))
```
Then, just use the userAtom in components.
If you'd like to create a custom hook, just `const useBrowserInfo = () => useAtom(userAtom)`.

https://jotai.org/docs/integrations/query

**Reply by amirhhashemi** [reply]
Thank you for the tip. I updated my question.

**Reply by t6adev** [reply]
> Is there any particular reason to prefer the second approach over the first one? I mean, they are both equal in terms of functionality, right?

IMO, there is no difference between them in this simple case.
But, as we write `const useBrowserInfo = () => useAtom(browserAtom)`, we don't need to connect them via hooks at all. We can put all logic into atoms without hooks.
Let's say if we put more params, like browserIdAtom, we have to add each line like `const [browserId] = useAtom(browserIdAtom) `.
If we take the second pattern, we can just put get(fooAtom). Very simple and straightforward.
It's a very important thing because if we want to write more complex logic in hooks, we have to pay attention to handling hooks rules.

**Reply by amirhhashemi** [reply]
You are right. I use several atoms in the custom hook and using `useAtom` to get each one's value is ugly:
```ts
const browserIdAtom = atomWithStorage("browserId", undefined)
const xAtom = atom("x", undefined)
const yAtom = atom("y", undefined)

export const useBrowserInfo = () => {
  const [browserId] = useAtom(browserIdAtom)
  const [x] = useAtom(xAtom)
  const [y] = useAtom(yAtom)

  // Other stuff...

  return useQuery({
    queryKey: ["browser", browserId, x, y], 
    queryFn: () => fetch(`/browser/${browserId}/${x}/${y}`)
}
```
Using the `get` function is a lot cleaner and I don't have to worry about hook rules:
```ts
const browserIdAtom = atomWithStorage("browserId", undefined)
const xAtom = atom("x", undefined)
const yAtom = atom("y", undefined)

const [browserAtom, browserStatusAtom] = atomsWithQuery((get) => ({
    queryKey: ["browser", get(browserIdAtom), get(xAtom), get(yAtom)], 
    queryFn: ({ queryKey: [, id, x, y] }) => fetch(`/browser/${id}/${x}/${y}`)
}))
```
I probably end up using `jotai-tanstack-query`. 

Thank you for the help

**Comment by AjaxSolutions** [comment]
Your example is rather simple.  You can do it all with atoms.  A jotai atom can fetch data whenever browserAtom or browserIdAtom is changed.  You can also use the loadable utility.

```
const browserDataAtom = atom(async (get) => {
  const browser = get(browserAtom);
  const browserId = get(browserIdAtom);

  const response = await fetch(`/browser/${browserId}`);
  const data = await response.json();

  return data;
});
```

**Reply by amirhhashemi** [reply]
I need some of the `react-query` features like `refetch-on-window-focus` and caching so I can't use `loadable` here. I over simplified my example. But thanks, I didn't know about `loadable`
